File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        longlen = 0
        
        def dfs(root):
            nonlocal longlen
            if not root:
                return 0

            if not root.left and not root.right:
                return 1
            
            leftheight = dfs(root.left)
            rightheight = dfs(root.right)
            curlen = leftheight + rightheight

            if curlen > longlen:
                longlen = curlen
            
            return max(leftheight, rightheight) + 1
        
        dfs(root)
        return longlen

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-0.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, 1);

        for (int i = nums.size() - 1; i >= 0; i--) {
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[i] < nums[j]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }

        return *max_element(dp.begin(), dp.end());
    }
};
